src/lsstseries/analysis/structure_function/iqr/calculator.py

```diff
@@ -0,0 +1,123 @@
+from typing import List
+
+import numpy as np
+from scipy.stats import binned_statistic
+
+from lsstseries.analysis.structure_function.base_argument_container import StructureFunctionArgumentContainer
+from lsstseries.analysis.structure_function.base_calculator import StructureFunctionCalculator
+
+# For details see Kozlowski 16 Equation 10: https://arxiv.org/abs/1604.05858
+CONVERSION_TO_SIGMA = 0.741
+
+
+class IqrStructureFunctionCalculator(StructureFunctionCalculator):
+    """This class implements the structure function calculation described in
+    Equation 10 of Kozlowski 16: https://arxiv.org/abs/1604.05858
+
+    SF_obs(deltaT) = 0.741 * IQR
+
+    Where `IQR` is the interquartile range between 25% and 75% of the sorted
+    (y(t) - y(t+delta_t)) distribution.
+    """
+
+    def __init__(
+        self,
+        time: List[List[float]],
+        flux: List[List[float]],
+        err: List[List[float]],
+        argument_container: StructureFunctionArgumentContainer,
+    ):
+        super().__init__(time, flux, err, argument_container)
+
+    def calculate(self):
+        all_d_fluxes = []
+        for lc_idx in range(len(self._time)):
+            lc_times = self._time[lc_idx]
+            lc_fluxes = self._flux[lc_idx]
+
+            # mask out any nan values
+            t_mask = np.isnan(lc_times)
+            f_mask = np.isnan(lc_fluxes)
+            lc_mask = np.logical_or(t_mask, f_mask)
+
+            lc_times = lc_times[~lc_mask]
+            lc_fluxes = lc_fluxes[~lc_mask]
+
+            # compute difference of times
+            dt_matrix = lc_times.reshape((1, lc_times.size)) - lc_times.reshape((lc_times.size, 1))
+            d_times = dt_matrix[dt_matrix > 0].flatten()
```

Comment:
In the paper you link, they define delta_t as abs(t_i - t_j). And then use the different pairings and a scale factor equal to the number of pairings. I think this is *mostly* (but not quite exactly) the same because you are throwing out approximately half the data (all the negative), which will effect the list of pairs and the number of pairings.

I'm not sure about the case of dt_matrix = 0 though. From a quick scan through the paper I'm not sure if the t_0 - t_0 is included in the pairings. If so that would be a data point for the IQR that you are dropping and could shift the boundaries a bit.

There is also the implicit assumption that there is only one observation at the first time. The pairings will be off if we have multiple observations at t_0 because they will all be filtered.

Reply:
Hmm, this is a good question. I'll have to ask Neven about this one. However, I can see that this code process is similar (identical) to what is show in this notebook: https://github.com/ywx649999311/structure_function/blob/master/notebooks/SF.ipynb that Neven shared earlier. 

But that being said, I'll spend some time with Neven to make sure that we're correctly interpreting what's in the paper. 

As another point of reference, we're using this paper as the source for several other SF methods https://arxiv.org/pdf/1401.1785.pdf and it's a little more clear about what constitutes a delta_t. 

"where ∆mij is the magnitude difference between the ith and jth observations (i < j) whose time lag (tj − ti) falls in the bin ∆t and with measurement uncertainties of σi and σj respectively"

Reply:
Just finished up a chat with Neven on this one. The approach here is considered correct. We're considering the time difference between observations at t_i and t_j where i < j. And you're correct, if there are multiple observations at the same time, we would not include the delta_magnitudes between those observations. 

Depending on the granularity of the timestamps, it might be a very rare case when two observations would happen at the same time.  If the finest timescale is a day, then this might be more common.